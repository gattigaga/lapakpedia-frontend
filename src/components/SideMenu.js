import React, { Component, createRef } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { Link } from "react-router-dom";
import { TweenLite, TimelineLite, Power4 } from "gsap";

const Container = styled.div`
  width: 280px;
  height: 100vh;
  background: #ddd;
  padding: 48px;
  box-sizing: border-box;
  position: fixed;
  left: -100%;
`;

const List = styled.ul`
  padding: 0px;
  margin: 0px;
`;

const ListItem = styled.li`
  list-style: none;
  transition: all 0.3s ease-out;
  margin-bottom: 24px;
`;

const ListCaption = styled(Link)`
  font-family: Roboto;
  font-size: 18px;
  color: black;
  text-decoration: none;
  letter-spacing: 0.1em;
`;

class SideMenu extends Component {
  constructor(props) {
    super(props);

    this.container = createRef();
    this.items = props.items.map(createRef);
  }

  componentWillReceiveProps(nextProps) {
    const { isOpen } = this.props;

    if (isOpen !== nextProps.isOpen) {
      const width = 280;
      const timeline = new TimelineLite();
      const items = this.items.map(item => item.current);

      timeline
        .fromTo(
          this.container.current,
          0.6,
          {
            ease: Power4.easeIn,
            left: nextProps.isOpen ? -width : 0
          },
          {
            ease: Power4.easeIn,
            left: nextProps.isOpen ? 0 : -width
          }
        )
        .staggerFromTo(
          items,
          0.2,
          {
            ease: Power4.easeInOut,
            x: nextProps.isOpen ? -32 : 0,
            opacity: nextProps.isOpen ? 0 : 1
          },
          {
            ease: Power4.easeInOut,
            x: nextProps.isOpen ? 0 : -32,
            opacity: nextProps.isOpen ? 1 : 0
          },
          0.1
        );
    }
  }

  render() {
    const { items } = this.props;

    return (
      <Container innerRef={this.container}>
        <List>
          {items.map((item, index) => (
            <ListItem key={index} innerRef={this.items[index]}>
              <ListCaption to={item.url}>{item.label}</ListCaption>
            </ListItem>
          ))}
        </List>
      </Container>
    );
  }
}

SideMenu.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      url: PropTypes.string
    })
  ),
  isOpen: PropTypes.bool
};

export default SideMenu;
