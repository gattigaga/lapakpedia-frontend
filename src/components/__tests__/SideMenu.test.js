import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import SideMenu from "components/SideMenu";

describe("SideMenu", () => {
  const setup = propOverrides => {
    const props = {
      items: [
        {
          label: "Home",
          url: "/home"
        },
        {
          label: "About",
          url: "/about"
        }
      ],
      ...propOverrides
    };

    const wrapper = shallow(<SideMenu {...props} />);

    return {
      props,
      wrapper
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
