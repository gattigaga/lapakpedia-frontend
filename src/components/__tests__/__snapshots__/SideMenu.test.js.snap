// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu should renders default 1`] = `
<styled.div
  innerRef={
    Object {
      "current": null,
    }
  }
>
  <styled.ul>
    <styled.li
      innerRef={
        Object {
          "current": null,
        }
      }
      key="0"
    >
      <Styled(Link)
        to="/home"
      >
        Home
      </Styled(Link)>
    </styled.li>
    <styled.li
      innerRef={
        Object {
          "current": null,
        }
      }
      key="1"
    >
      <Styled(Link)
        to="/about"
      >
        About
      </Styled(Link)>
    </styled.li>
  </styled.ul>
</styled.div>
`;
